### ISSUE TITLE
Fix hacky approach for detecting whether a widget is inactive

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2612]
As part of the dynamic widgets API, we want to make it simple for developers to implement their widgets without calling specific Site Kit datastore actions to indicate a certain state of the widgets. #2252 focused on introducing special `Widget*` wrapper components for various special states so that the registered widget component simply needs to return that component. This is further simplified since these components are passed to the widget component as props (in a scoped version), so developers don't need to import it.

For a couple months now, we've had some _terribly_ hacky and ugly logic in `WidgetAreaRenderer` which inspects the return value of the registered widget component by calling its render function directly - while this works, it goes against React best practices, where overall inspecting child components in such a way is not intended. This hack furthermore causes some issues which (among others) prevent us from updating to a newer version of React.

We should replace this hack with a proper mechanism to detect whether a widget is inactive. One possible approach would be to follow along the pattern that #2252 introduced:
* We could introduce a `WidgetNull` component and pass it to every registered widget component.
* The component would dispatch a new `setWidgetActive( false )` action on mount and `setWidgetActive( true )` on unmount, similar to how e.g. `WidgetReportZero` sets widget state. The component would technically just return `null`, i.e. render nothing.
* `WidgetAreaRenderer` could then recognize from the datastore for each widget whether it is active or inactive.
* This would require registered components to return `<WidgetNull />` instead of `null`, which maybe is a bit unintuitive. But it could still be a good solution, we can always document it.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* There should be a new `Null` component in `assets/js/components` which simply returns `null`.
* There should be a new `WidgetNull` component in `assets/js/googlesitekit/widgets/components` that requires a `widgetSlug` prop.
    * It should include a `useEffect` hook that calls `setWidgetState( widgetSlug, Null, {} )`, and `unsetWidgetState` with the same arguments as cleanup function. (This is similar to e.g. `WidgetActivateModuleCTA` or `WidgetReportZero`.)
    * It should render `<Null />`.
* The `getWidgetComponentProps` function should be expanded to also include a scoped version of `WidgetNull` in the returned object (which is then passed to every widget's component).
* There should be a new function `separateNullWidgets( widgets, widgetStates )` in `assets/js/googlesitekit/widgets/util`:
    * Given an array of widget objects and a map of widget slugs and their states (one entry per widget), it should check for each widget whether its state is a "null" state (i.e. the `Component` is `Null`).
    * It should based on that check split the given `widgets` into two arrays, one `activeWidgets` and the other `inactiveWidgets`.
    * It should then return the object `{ activeWidgets, inactiveWidgets }`.
* The `WidgetAreaRenderer` should be updated:
    * The current hacky filtering function that assigns `activeWidgets` should be replaced with calling `separateNullWidgets( widgets, widgetStates )`.
    * The `activeWidgets` returned should go through the same remaining component logic that they currently do.
    * The `inactiveWidgets` should simply be rendered within a new `.googlesitekit-widget-area-widgets > .googlesitekit-hidden` container, i.e. hidden via CSS. (All of these inactive widgets should return `null` anyway so the extra container and hiding it via CSS might seem unnecessary at first glance, but for the split second in which a widget might e.g. become active again after being inactive, it's safer to render in a hidden `div`.)
* All widget components (i.e. components used with the JS Widgets API) should be modified to return `<WidgetNull />` received from props instead of `null`. Basically, no widget should ever return just `null`, it should always be via `WidgetNull`.

## Implementation Brief

* Implement new components `Null` and `WidgetNull` per the ACs, following the similar approach from e.g. `WidgetActivateModuleCTA` (a lot can be copy-pasted from there).
* Include scoped `WidgetNull` component where the `widgetSlug` prop is already provided in `getWidgetComponentProps`.
* Implement new `separateNullWidgets` utility function. Use `widgets.reduce( widgets, ( separated, widget ) => { ... }, { activeWidgets: [], inactiveWidgets: [] } )` to compute the return value.
* Update `WidgetAreaRenderer` to use `separateNullWidgets` per the ACs.
* Update the `whenActive` HOC to return `<WidgetNull />` instead of `null`. (All our widgets rely on that for the only scenarios where widgets return `null`, so there is no need to update any widget component directly.)

### Test Coverage

* Add Jest test coverage for the `WidgetNull` component.
* Add Jest test coverage for the `separateNullWidgets` function.

### Visual Regression Changes

* N/A

## QA Brief

New JS tests should pass

WidgetNull
https://github.com/google/site-kit-wp/blob/develop/assets/js/googlesitekit/widgets/components/WidgetNull.test.js

get-widget-layout
https://github.com/google/site-kit-wp/blob/develop/assets/js/googlesitekit/widgets/util/get-widget-layout.test.js

isInactiveWidgetState
https://github.com/google/site-kit-wp/blob/develop/assets/js/googlesitekit/widgets/util/is-inactive-widget-state.test.js

combine-widgets
https://github.com/google/site-kit-wp/blob/develop/assets/js/googlesitekit/widgets/util/combine-widgets.test.js

The dashboard page should render the same before and after the change.
/wp-admin/admin.php?page=googlesitekit-dashboard

To manually inspect the functionality, a breakpoint here will assist with that https://github.com/google/site-kit-wp/blob/e3bf45b2b2e06227505c77350e21b3fecfd482b4/assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.js#L41

Something like the following will assist getting to the right state.
```js	
if ( JSON.stringify( activeWidgets ) !== JSON.stringify( widgets ) ) {
  debugger; // eslint-disable-line no-debugger
}
```

The widget component for an inactive widget will be WidgetNull, `activeWidgets` will differ in content from `widgets`. An inactive widget will render nothing as WidgetNull renders Null component which returns null.

## Changelog entry

* Use reliable and React-friendly approach for detecting whether a widget is inactive and should not be rendered.

**Comment by aaemnnosttv** [comment]
@felixarntz I thought about doing the same thing as well and introducing a `WidgetNull` component is probably the best option I can think of at the moment.

Another option would be to expose a registration option that would take a function, similar to `validateSubmitChanges` which would return a boolean whether or not the widget would render or not. This would probably work too and could even use `WidgetNull` internally in combination with one of the higher widget components.

I think introducing a `WidgetNull` and passing that as a prop like the others is the simplest change and fits with our convention that anything rendered (including nothing) other than the widget's own output should come from a prop-provided component type.

**Comment by aaemnnosttv** [comment]
Let's make sure all of the IB sections are always filled out, even if no changes are required to differentiate between no action needed and an incomplete IB. Regarding the VRT section, I think it's safe to say there would be no changes here so I'll add that quickly to move this along.

IB ✅

**Comment by ivankruchkoff** [comment]
Perhaps I'm missing something, but which part handles this part of the task description:

The component would dispatch a new `setWidgetActive( false )` action on mount and `setWidgetActive( true )` on unmount, similar to how e.g. WidgetReportZero sets widget state. The component would technically just return null, i.e. render nothing.

This is what's missing to get this into CR as remaining functionality is in place.

**Comment by felixarntz** [comment]
@ivankruchkoff It looks like the original description was bit too close to code and was revised in the ACs; we eventually decided to use `setWidgetState` and `unsetWidgetState` instead of a new `setWidgetActive` action. Essentially:
* A call to `setWidgetState( widgetSlug, Null, {} )` is the same as `setWidgetActive( widgetSlug, false )` would have been.
* A call to `unsetWidgetState( widgetSlug, Null, {} )` is the same as `setWidgetActive( widgetSlug, true )` would have been.

Basically, by setting the `Null` component as widget state, we indicate that the widget state is inactive. You can ignore the part about the `setWidgetActive` action, the ACs have the intended approach.

**Comment by ivankruchkoff** [comment]
Over to CR.

@felixarntz this approach takes a slightly different path to the AC/IB, @aaemnnosttv and I have discussed it quite a bit recently. There were some issues with infinite renders occurring, and this somewhat cleans it up.

Here are some of those differences:
1. Removed separateNullWidgets, instead there is  selector isWidgetActive - https://github.com/google/site-kit-wp/pull/2731/files#diff-0e8bdf4edf410f93711eb977fe477d9ac392f727af1807719e5b3f4dbcbac606R321
2. All widgets are rendered, and the logic for rendering the widget vs WidgetNull occurs in WidgetRenderer and not WidgetAreaRenderer https://github.com/google/site-kit-wp/pull/2731/files#diff-0c87147dda5b3ec1aaf55daab22d497837989e410fa0a3d4dbd632bccbde876aR42
3. Inactive widgets (that should be widgetnull), also get a display: none css property.

Based on 2 and 3, one widget rendering test had to be updated as the widget is rendered but with display none attribute, see https://github.com/google/site-kit-wp/blob/b909660280d51ac8179ca38ae91511de6ea3ce38/assets/js/googlesitekit/widgets/components/__snapshots__/WidgetAreaRenderer.test.js.snap

**Comment by felixarntz** [comment]
@ivankruchkoff Can you add a QA Brief to this one? It can be relatively engineer-focused since the changes here are mostly of technical nature and they shouldn't have a user-facing impact.

**Comment by tofumatt** [comment]
QA ✅

All the steps in the QA brief checked out, thanks for the easy debugger line 😄
